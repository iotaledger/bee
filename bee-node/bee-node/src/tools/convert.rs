// Copyright 2020-2022 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use bee_block::address::{Address, Ed25519Address};
use crypto::hashes::{blake2b::Blake2b256, Digest};
use structopt::StructOpt;
use thiserror::Error;

#[derive(Clone, Debug, Error)]
pub enum ConvertError {
    #[error("invalid Bech32 address length")]
    InvalidAddressLength(),
    #[error("Invalid Address")]
    InvalidAddress(),
}

#[derive(Clone, Debug, StructOpt)]
pub enum ConvertTool {
    /// Converts a Bech32 address to a hex encoded one.
    Bech32ToHex {
        #[structopt(long)]
        bech32: String,
    },
    /// Converts a hex encoded address to a Bech32 one.
    HexToBech32 {
        #[structopt(long)]
        hex: String,
        #[structopt(long)]
        hrp: String,
    },
    /// Converts a hex encoded public key to a Bech32 address.
    HexPubkeyToBech32 {
        #[structopt(long)]
        pubkey: String,
        #[structopt(long)]
        hrp: String,
    },
}

pub fn exec(tool: &ConvertTool) -> Result<(), ConvertError> {
    match tool {
        ConvertTool::Bech32ToHex { bech32 } => {
            let hex = bech32_to_hex(bech32.as_str());
            match hex {
                Ok(_) => println!("Your Hex encoded address is:\t{}", hex.unwrap()),
                Err(e) => println!("Error: {}", e),
            }
        }
        ConvertTool::HexToBech32 { hex, hrp } => {
            let bech32 = hex_to_bech32(hex.as_str(), hrp.as_str());
            match bech32 {
                Ok(_) => println!("Your Bech32 address is:\t{:?}", bech32.unwrap()),
                Err(e) => println!("Error: {}", e),
            }
        }
        ConvertTool::HexPubkeyToBech32 { pubkey, hrp } => {
            let bech32 = hex_public_key_to_bech32_address(pubkey.as_str(), hrp.as_str());
            match bech32 {
                Ok(_) => println!("Your Bech32 address is:\t{:?}", bech32.unwrap()),
                Err(e) => println!("Error: {}", e),
            }
        }
    }
    Ok(())
}

/// Transforms bech32 to hex
fn bech32_to_hex(bech32: &str) -> Result<String, ConvertError> {
    let address = Address::try_from_bech32(bech32).map_err(|_| ConvertError::InvalidAddress())?;
    let hex_string = match address {
        (_, Address::Ed25519(ed)) => ed.to_string(),
        (_, Address::Alias(alias)) => alias.to_string(),
        (_, Address::Nft(nft)) => nft.to_string(),
    };
    Ok(hex_string)
}

/// Transforms a hex encoded address to a bech32 encoded address
fn hex_to_bech32(hex: &str, bech32_hrp: &str) -> Result<String, ConvertError> {
    let address: Ed25519Address = hex
        .parse::<Ed25519Address>()
        .map_err(|_| ConvertError::InvalidAddress())?;
    Ok(Address::Ed25519(address).to_bech32(bech32_hrp))
}

/// Transforms a hex encoded public key to a bech32 encoded address
fn hex_public_key_to_bech32_address(hex: &str, bech32_hrp: &str) -> Result<String, ConvertError> {
    let mut public_key = [0u8; Ed25519Address::LENGTH];
    hex::decode_to_slice(&hex, &mut public_key).map_err(|_| ConvertError::InvalidAddressLength())?;

    let address = Blake2b256::digest(&public_key)
        .try_into()
        .map_err(|_e| ConvertError::InvalidAddress())?;
    let address: Ed25519Address = Ed25519Address::new(address);
    Ok(Address::Ed25519(address).to_bech32(bech32_hrp))
}

#[cfg(test)]
mod bech32tests {
    use crate::tools::convert::*;
    // spec: https://github.com/iotaledger/tips/blob/main/tips/TIP-0011/tip-0011.md
    const HRP: &str = "iota";

    #[test]
    fn test_bech32_to_hex() {
        let bech32_address = "iota1qrhacyfwlcnzkvzteumekfkrrwks98mpdm37cj4xx3drvmjvnep6xqgyzyx".to_string();
        let hex_encoded_address = bech32_to_hex(&bech32_address).unwrap();

        assert_eq!(
            &hex_encoded_address,
            "0xefdc112efe262b304bcf379b26c31bad029f616ee3ec4aa6345a366e4c9e43a3"
        );

        let bech32_to_hex = ConvertTool::Bech32ToHex { bech32: bech32_address };
        exec(&bech32_to_hex).unwrap();
    }

    #[test]
    fn test_hex_to_bech32() {
        let hex_address = "0xefdc112efe262b304bcf379b26c31bad029f616ee3ec4aa6345a366e4c9e43a3".to_string();
        let bech32_address = hex_to_bech32(&hex_address, HRP).unwrap();

        assert_eq!(
            &bech32_address,
            "iota1qrhacyfwlcnzkvzteumekfkrrwks98mpdm37cj4xx3drvmjvnep6xqgyzyx"
        );

        let hex_to_bech32 = ConvertTool::HexToBech32 {
            hex: hex_address,
            hrp: HRP.to_string(),
        };
        exec(&hex_to_bech32).unwrap();
    }

    #[test]
    fn test_public_key_to_bech32() {
        let public_key = "6f1581709bb7b1ef030d210db18e3b0ba1c776fba65d8cdaad05415142d189f8".to_string();
        let bech32_address = hex_public_key_to_bech32_address(&public_key, HRP).unwrap();

        assert_eq!(
            &bech32_address,
            "iota1qrhacyfwlcnzkvzteumekfkrrwks98mpdm37cj4xx3drvmjvnep6xqgyzyx"
        );

        let public_key_to_bech32 = ConvertTool::HexPubkeyToBech32 {
            pubkey: public_key,
            hrp: HRP.to_string(),
        };
        exec(&public_key_to_bech32).unwrap();
    }
}
