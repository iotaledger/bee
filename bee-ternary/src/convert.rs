// Copyright 2020 IOTA Stiftung
//
// Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
// the License. You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
// an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and limitations under the License.

use crate::{Btrit, RawEncoding, RawEncodingBuf, Trit, TritBuf, Trits, Utrit};
use num_traits::{AsPrimitive, CheckedAdd, CheckedSub, FromPrimitive, Num, Signed};
use std::{cmp::PartialOrd, convert::TryFrom};

/// An error that may be produced during numeric conversion.
#[derive(Debug, PartialEq)]
pub enum Error {
    /// The trit slice didn't contain enough trits to be considered a numeric value.
    Empty,
    /// An overflow occurred during a numeric operation.
    Overflow,
    /// An underflow occurred during a numeric operation.
    Underflow,
}

// TODO: Find a way to implement this without conflicting impls
// impl<I, T: RawEncodingBuf> From<I> for TritBuf<T>
// where
//     T::Slice: RawEncoding<Trit = Btrit>,
//     I: AsPrimitive<i8> + FromPrimitive + Signed,
// {
//     fn from(x: I) -> Self {
//         signed_int_trits(x).collect()
//     }
// }

macro_rules! signed_try_from_trits {
    ($int:ident) => {
        impl<'a, T: RawEncoding<Trit = Btrit> + ?Sized> TryFrom<&'a Trits<T>> for $int {
            type Error = Error;
            fn try_from(trits: &'a Trits<T>) -> Result<Self, Self::Error> {
                trits_to_int(trits)
            }
        }

        impl<T: RawEncodingBuf> From<$int> for TritBuf<T>
        where
            T::Slice: RawEncoding<Trit = Btrit>,
        {
            fn from(x: $int) -> Self {
                signed_int_trits(x).collect()
            }
        }
    };
}

// We have to implement manually due to Rust's orphan rules :(
// This macro accepts anything that implements:
// `Clone + CheckedAdd + Signed + AsPrimitive<i8> + FromPrimitive`
#[cfg(has_i128)]
signed_try_from_trits!(i128);
signed_try_from_trits!(i64);
signed_try_from_trits!(i32);
signed_try_from_trits!(i16);
signed_try_from_trits!(i8);

macro_rules! unsigned_try_from_trits {
    ($int:ident) => {
        impl<'a, T: RawEncoding<Trit = Utrit> + ?Sized> TryFrom<&'a Trits<T>> for $int {
            type Error = Error;
            fn try_from(trits: &'a Trits<T>) -> Result<Self, Self::Error> {
                trits_to_int(trits)
            }
        }

        impl<T: RawEncodingBuf> From<$int> for TritBuf<T>
        where
            T::Slice: RawEncoding<Trit = Utrit>,
        {
            fn from(x: $int) -> Self {
                unsigned_int_trits(x).collect()
            }
        }
    };
}

// We have to implement manually due to Rust's orphan rules :(
// This macro accepts anything that implements:
// `Clone + CheckedAdd + Unsigned + AsPrimitive<u8> + FromPrimitive`
#[cfg(has_i128)]
unsigned_try_from_trits!(u128);
unsigned_try_from_trits!(u64);
unsigned_try_from_trits!(u32);
unsigned_try_from_trits!(u16);
unsigned_try_from_trits!(u8);

/// Attempt to convert the given trit slice into a number. If the numeric representation of the
/// trit slice is too large or small to fit the numeric type, or does not contain any trits, an
/// error will be returned.
pub fn trits_to_int<I, T: RawEncoding + ?Sized>(trits: &Trits<T>) -> Result<I, Error>
where
    I: Clone + CheckedAdd + CheckedSub + PartialOrd + Num,
{
    if trits.is_empty() {
        Err(Error::Empty)
    } else {
        let mut acc = I::zero();

        for trit in trits.iter().rev() {
            let old_acc = acc.clone();
            acc = trit
                .add_to_num(acc)?
                .checked_add(&old_acc)
                .and_then(|acc| acc.checked_add(&old_acc))
                .ok_or_else(|| {
                    if old_acc < I::zero() {
                        Error::Underflow
                    } else {
                        Error::Overflow
                    }
                })?;
        }

        Ok(acc)
    }
}

/// Produce an iterator over the [`Btrit`]s that make up a given integer.
pub fn signed_int_trits<I>(x: I) -> impl Iterator<Item = Btrit> + Clone
where
    I: Clone + AsPrimitive<i8> + FromPrimitive + Signed,
{
    let is_neg = x.is_negative();
    let mut x = if is_neg { x } else { -x };

    let radix = I::from_i8(3).unwrap();

    std::iter::from_fn(move || {
        if x.is_zero() {
            None
        } else {
            let modulus = ((x + I::one()).abs() % radix).as_();
            x = x / radix;
            if modulus == 1 {
                x = x + -I::one();
            }
            Some(Btrit::try_from(((modulus + 2) % 3 - 1) * if is_neg { -1 } else { 1 }).unwrap())
        }
    })
    // If the integer is exactly 0, add an extra trit
    .chain(Some(Btrit::Zero).filter(|_| x.is_zero()))
}

/// Produce an iterator over the [`Utrit`]s that make up a given integer.
pub fn unsigned_int_trits<I>(mut x: I) -> impl Iterator<Item = Utrit> + Clone
where
    I: Clone + AsPrimitive<u8> + FromPrimitive + Num,
{
    let radix = I::from_u8(3).unwrap();

    std::iter::from_fn(move || {
        if x.is_zero() {
            None
        } else {
            let modulus = (x % radix).as_();
            x = x / radix;
            Some(Utrit::try_from(modulus).unwrap())
        }
    })
    // If the integer is exactly 0, add an extra trit
    .chain(Some(Utrit::Zero).filter(|_| x.is_zero()))
}
