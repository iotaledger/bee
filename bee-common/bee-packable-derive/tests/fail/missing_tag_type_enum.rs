// Copyright 2021 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

#![allow(unused_imports)]

use bee_packable::Packable;

use core::convert::Infallible;

#[derive(Packable)]
#[packable(pack_error = Infallible)]
#[packable(unpack_error = Infallible)]
pub enum OptI32 {
    #[packable(tag = 0)]
    None,
    #[packable(tag = 1)]
    Some(i32),
}

fn main() {}
