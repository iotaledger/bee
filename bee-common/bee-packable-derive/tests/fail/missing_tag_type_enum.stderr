error: Enums that derive `Packable` require a `#[packable(tag_type = ...)]` or `#[repr(...)]` attribute.
  --> tests/fail/missing_tag_type_enum.rs:12:10
   |
12 | pub enum OptI32 {
   |          ^^^^^^
